ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -162,52 +216,83 @@ densityDisconnect ::
+          -- horizon that we already received), there can be no headers in between and 'potentialSlots' is 0.
+          SlotNo potentialSlots
+            | caughtUp
+            = 0
             | otherwise
-            = AF.length candidateSuffix >= fromIntegral k
-          lastSlot | isJust maybeForecast, lastSlotForecast >= lastSlotCandidate = Left lastSlotForecast
-                   | otherwise = Right lastSlotCandidate
-      pure (peer, DensityBounds {fragment, offersMoreThanK, lowerBound, upperBound, hasBlockAfter = hasBlockAfterGenesisWindow, lastSlot})
+            = case latestSlot of
+              -- The peer either has not advertised its tip yet or simply has no blocks whatsoever and won't progress
+              -- either.
+              -- In the latter case, it should be killed by the LoP.
+              -- REVIEW: What do we want to do in tests here? Both cases are possible, and we cannot distinguish them
+              -- without the tip (even with the tip we have been relying on the invariant that each peer advertises the
+              -- tip before any of them sends headers).
+              -- NoLatestSlot -> 0
+              NoLatestSlot -> SlotNo sgen
+              -- 'endOfGenesisWindow' is exclusive, so we have to add 1 to the non-exclusive last slot
+              LatestSlotCandidate slot -> intervalLength endOfGenesisWindow (slot + 1)
+              -- If the candidate fragment's last slot is smaller than the slot set after receiving a header, we are
+              -- stuck at the forecast horizon, which implies that there is a header after the Genesis window.
+              LatestSlotForecast _ -> 0
+
+          -- The number of blocks within the Genesis window we know with certainty
+          lowerBound = fromIntegral $ AF.length fragment
+
+          upperBound = lowerBound + potentialSlots
+
+          -- The number of blocks we know of on the candidate chain, not limited to the Genesis window.
+          totalBlockCount = fromIntegral (AF.length candidateSuffix)
+
+          -- Does the peer have more than k known blocks in _total_ after the intersection?
+          -- If not, it is not qualified to compete by density (yet).
+          offersMoreThanK = case latestSlot of
+            NoLatestSlot          -> False
+            LatestSlotCandidate _ -> totalBlockCount > k
+            -- If the last slot is not on the candidate chain, we know that there is at least one more block, so we use
+            -- '>=' here.
+            LatestSlotForecast _  -> totalBlockCount >= k
+
+          -- For tracing: Is there a block after the end of the Genesis window?
+          -- Note that 'endOfGenesisWindow' is exclusive, so we use '>='.
+          hasBlockAfter = case latestSlot of
+            NoLatestSlot             -> False
+            LatestSlotCandidate slot -> slot >= endOfGenesisWindow
+            LatestSlotForecast slot  -> slot >= endOfGenesisWindow
```

Comment:
`endOfGenesisWindow` points at the first slot after the genesis window. I feel like this would deserve a renaming to `firstSlotAfterGenesisWindow`. (`endOfGenesisWindow` means to me the last slot of the genesis window, i.e. `intersection_slot + sgen`).

Reply:
Renamed.